File: fastapi/agent_profile.py
"""Fixed-path Agent behavior profile shared with the Node template."""

from __future__ import annotations

import json
from pathlib import Path

DEFAULT_AGENT_DIRECTORY = Path(__file__).resolve().parents[1] / "agent"
MAX_PROFILE_BYTES = 4_096
MAX_INSTRUCTIONS_BYTES = 32_768
PROFILE_KEYS = {"display_name", "id", "instructions_file", "schema"}


def _read_fixed_file(path: Path, maximum: int, label: str, root: Path) -> str:
    resolved_root = root.resolve(strict=True)
    if path.is_symlink() or not path.is_file():
        raise ValueError(f"{label} must be a regular non-symlink file")
    resolved = path.resolve(strict=True)
    if resolved != resolved_root and resolved_root not in resolved.parents:
        raise ValueError(f"{label} must stay inside the fixed agent directory")
    size = path.stat().st_size
    if not 1 <= size <= maximum:
        raise ValueError(f"{label} must contain 1 to {maximum} bytes")
    try:
        return path.read_bytes().decode("utf-8", errors="strict")
    except UnicodeDecodeError as exc:
        raise ValueError(f"{label} must be valid UTF-8") from exc


def load_agent_profile(agent_directory: Path | None = None) -> dict[str, str]:
    root = (agent_directory or DEFAULT_AGENT_DIRECTORY).resolve()
    profile_path = root / "profile.json"
    try:
        profile = json.loads(
            _read_fixed_file(profile_path, MAX_PROFILE_BYTES, "agent profile", root)
        )
    except json.JSONDecodeError as exc:
        raise ValueError("agent profile must be valid JSON") from exc
    if not isinstance(profile, dict) or set(profile) != PROFILE_KEYS:
        raise ValueError(
            "agent profile must contain exactly: "
            + ", ".join(sorted(PROFILE_KEYS))
        )
    if profile["schema"] != "sage-agent-profile-v1":
        raise ValueError("unsupported agent profile schema")
    identifier = profile["id"]
    if (
        not isinstance(identifier, str)
        or not 3 <= len(identifier) <= 64
        or not identifier[0].isalpha()
        or not identifier[0].islower()
        or any(character not in "abcdefghijklmnopqrstuvwxyz0123456789-" for character in identifier)
    ):
        raise ValueError("agent profile id must be a bounded lowercase slug")
    display_name = profile["display_name"]
    if (
        not isinstance(display_name, str)
        or display_name != display_name.strip()
        or not 1 <= len(display_name) <= 120
    ):
        raise ValueError("agent profile display_name must contain 1 to 120 characters")
    if profile["instructions_file"] != "instructions.md":
        raise ValueError(
            "agent profile instructions_file must be exactly instructions.md"
        )
    instructions = _read_fixed_file(
        root / "instructions.md",
        MAX_INSTRUCTIONS_BYTES,
        "agent instructions",
        root,
    ).strip()
    if not instructions or "\x00" in instructions:
        raise ValueError(
            "agent instructions must contain bounded text without NUL bytes"
        )
    return {
        "display_name": display_name,
        "id": identifier,
        "instructions": instructions,
    }

File: node/agent-profile.mjs
import {
  lstatSync,
  readFileSync,
  realpathSync,
} from "node:fs";
import { dirname, resolve, sep } from "node:path";
import { fileURLToPath } from "node:url";

const DEFAULT_AGENT_DIRECTORY = fileURLToPath(new URL("../agent/", import.meta.url));
const MAX_PROFILE_BYTES = 4_096;
const MAX_INSTRUCTIONS_BYTES = 32_768;
const PROFILE_KEYS = ["display_name", "id", "instructions_file", "schema"];

function readFixedFile(path, maximum, label, root) {
  const absolute = resolve(path);
  const realRoot = realpathSync(root);
  const realPath = realpathSync(absolute);
  if (realPath !== realRoot && !realPath.startsWith(`${realRoot}${sep}`)) {
    throw new Error(`${label} must stay inside the fixed agent directory`);
  }
  const metadata = lstatSync(absolute);
  if (!metadata.isFile() || metadata.isSymbolicLink()) {
    throw new Error(`${label} must be a regular non-symlink file`);
  }
  if (metadata.size < 1 || metadata.size > maximum) {
    throw new Error(`${label} must contain 1 to ${maximum} bytes`);
  }
  const bytes = readFileSync(absolute);
  try {
    return new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  } catch {
    throw new Error(`${label} must be valid UTF-8`);
  }
}
function exactKeys(value, keys) {
  return value &&
    typeof value === "object" &&
    !Array.isArray(value) &&
    JSON.stringify(Object.keys(value).sort()) === JSON.stringify([...keys].sort());
}

export function loadAgentProfile(agentDirectory = DEFAULT_AGENT_DIRECTORY) {
  const root = resolve(agentDirectory);
  const profilePath = resolve(root, "profile.json");
  let profile;
  try {
    profile = JSON.parse(readFixedFile(profilePath, MAX_PROFILE_BYTES, "agent profile", root));
  } catch (error) {
    if (error instanceof SyntaxError) throw new Error("agent profile must be valid JSON");
    throw error;
  }
  if (!exactKeys(profile, PROFILE_KEYS)) {
    throw new Error(`agent profile must contain exactly: ${PROFILE_KEYS.join(", ")}`);
  }
  if (profile.schema !== "sage-agent-profile-v1") {
    throw new Error("unsupported agent profile schema");
  }
  if (typeof profile.id !== "string" || !/^[a-z][a-z0-9-]{2,63}$/.test(profile.id)) {
    throw new Error("agent profile id must be a bounded lowercase slug");
  }
  if (
    typeof profile.display_name !== "string" ||
    profile.display_name !== profile.display_name.trim() ||
    profile.display_name.length < 1 ||
    profile.display_name.length > 120
  ) {
    throw new Error("agent profile display_name must contain 1 to 120 characters");
  }
  if (profile.instructions_file !== "instructions.md") {
    throw new Error("agent profile instructions_file must be exactly instructions.md");
  }
  const instructionsPath = resolve(dirname(profilePath), "instructions.md");
  const instructions = readFixedFile(
    instructionsPath,
    MAX_INSTRUCTIONS_BYTES,
    "agent instructions",
    root,
  ).trim();
  if (!instructions || instructions.includes("\0")) {
    throw new Error("agent instructions must contain bounded text without NUL bytes");
  }
  return Object.freeze({
    displayName: profile.display_name,
    id: profile.id,
    instructions,
  });
}
